muffnn/autoencoder/autoencoder.py

```diff
@@ -0,0 +1,686 @@
+
+            # Anything left uses the default metric.
+            if self._default_indices is not None:
+                default_msk = tf.transpose(self._default_msk)
+
+                if self.metric == 'mse':
+                    if self.output_activation is not None:
+                        t = ((1.0 - default_msk) * t +
+                             default_msk * self.output_activation(t))
+                    tsub = tf.gather(t, self._default_indices)
+                    isub = tf.gather(tf.transpose(self._input_values),
+                                     self._default_indices)
+                    diff = isub - tsub
+                    scores += tf.reduce_sum(tf.square(diff),
+                                            reduction_indices=[0])
+                elif self.metric == 'cross-entropy':
+                    tsub = tf.gather(t, self._default_indices)
+                    isub = tf.gather(tf.transpose(self._input_values),
+                                     self._default_indices)
+
+                    if self.output_activation is tf.nn.sigmoid:
+                        scores += tf.reduce_sum(
+                            tf.nn.sigmoid_cross_entropy_with_logits(tsub,
+                                                                    isub),
+                            reduction_indices=[0])
+                        t = ((1.0 - default_msk) * t +
+                             default_msk * tf.nn.sigmoid(t))
+                    elif self.output_activation is tf.nn.softmax:
+                        # More pain here. The softmax has an implicit sum,
+                        # and since we cannot scatter into Tensors, we
+                        # break up the comp into the denominator and the
+                        # numerator using the mask.
+                        scores += tf.nn.softmax_cross_entropy_with_logits(
+                            tsub, isub, dim=0)
+                        log_softmax_denom = tf.reduce_logsumexp(
+                            tsub, reduction_indices=[0])
+                        t = ((1.0 - default_msk) * t +
+                             default_msk * tf.exp(t - log_softmax_denom))
+                    else:
+                        raise ValueError("Only `tensorflow.nn.sigmoid` and "
+                                         "`tensorflow.nn.softmax` output "
+                                         "activations can be used with the "
+                                         "'cross-entropy' metric!")
+                else:
+                    raise ValueError('Metric "%s" is not allowed!' %
+                                     self.metric)
+
+            # Undo the transpose here.
+            t = tf.transpose(t)
+
+        return t, scores
+
+    def _make_feed_dict(self, X, inverse=False, training=False):
+        # Make the dictionary mapping tensor placeholders to input data.
+
+        # Convert sparse inputs to dense.
+        if sp.issparse(X):
```

Comment:
Two things:

1. No, `check_array` will convert various input sparse formats to a single format, but does not do the sparse to dense conversion. 

2. My goal here was to only convert a sparse input to dense for each mini-batch instead of doing the whole data set at once in order to save memory. This is useful for giant bag-of-words representations, for example.